OSDOCS-3048 Tech preview of KEDA in 4.11

mburke5678 opened the pull request:
https://issues.redhat.com/browse/OSDOCS-3048

Preview: http://file.rdu.redhat.com/mburke/OSDOCS-3048-custom-autoscaling-keda/nodes/pods/nodes-pods-autoscaling-custom.html

**Release 7/25**

bergerhoffer commented:
The `branch/enterprise-4.11` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.10. And any PR going into `main` must also target the latest version branch (enterprise-4.11).

If the update in your PR does NOT apply to version 4.11 onward, please retarget this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

mburke5678 commented:
@joelsmith PTAL

joelsmith commented:
> You can create an CMA for any deployment, deployment config, replica set, replication controller, stateful set, or job.

Should be `a CMA` (not an).

> Your other applications continue to use other scaling methods.

I found this a little confusing. Do you mean to say that CMA will ignore all workloads which do not have a corresponding scaledobject/scaledjob?

> The CMA is better suited for workloads that are event-driven or uses specific meteric.

Problems with singular/plural, misspelled metric. Maybe you'll be going through a copy edit phase after the content is settled, so sorry if I'm jumping the gun on these edits.

The uninstall section should probably include instructions for uninstalling the KEDA OLM Operator via OperatorHub. I doubt that uninstalling the two CRDs will be sufficient. The KEDA uninstall process will be similar to the CMA uninstall process. I'll work on providing both to you.

I think that the output is out of date for this command: `$ oc get all -n openshift-custom-metrics-autoscaler`

You should see `custom-metrics-autoscaler-operator` instead of `keda-olm-operator`.

There appears to be some VPA copy/paste stuff that should probably be removed. You'll spot at least one if you search for VPA, and I also noticed this paragraph:

> You can remove the Custom Metrics Autoscaler Operator (CMA) from your OpenShift Container Platform cluster. After uninstalling, the resource requests for the pods already modified by an existing CMA CR do not change. Any new pods get the resources defined in the workload object, not the previous recommendations made by the Custom Metrics Autoscaler Operator.

I'm not sure what the behavior of an HPA would be if KEDA gets uninstalled and the HPA objects are left behind. I believe it that it would stop scaling.  The bit about resource requests and recommendations is a VPA thing. KEDA is only going to cause the number of replicas to change (via HPA), so anything that references resource requests is a vestige of the VPA docs.

I'll work on providing comprehensive uninstall instructions for you.

mburke5678 commented:
@weinliu PTAL

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger-auth.adoc:
```diff
@@ -0,0 +1,128 @@
+
+A trigger authentication allows you to include authentication information in a scaled object or a scaled job that can be used by the associated containers. You can use trigger authentications to pass {product-title} secrets, platform-native pod authentication mechanisms, environment variables, and so on.   
+
+Some triggers, also known as scalers, allow you to set specific authentication parameters by using a trigger authentication. For information, see each scaler/trigger definition in the link:https://keda.sh/docs/2.7/scalers/[Keda Scalers] documentation.
```
```suggestion
Some triggers, also known as scalers, allow you to set specific authentication parameters by using a trigger authentication. For information, see each scaler/trigger definition in the link:https://keda.sh/docs/2.7/scalers/[KEDA Scalers] documentation.
```

zroubalik commented on modules/nodes-pods-autoscaling-custom-creating-job.adoc:
```diff
@@ -0,0 +1,145 @@
+[id="nodes-pods-autoscaling-custom-creating-job_{context}"]
+= Adding a custom metrics autoscaler to a job
+
+You can create a custom metrics autoscaler (CMA) for any `Job` object.
```
ScaledJob Spec should be here: https://keda.sh/docs/2.7/concepts/scaling-jobs/#scaledjob-spec

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,157 @@
+
+[discrete]
+[id="nodes-pods-autoscaling-custom-trigger-prom_{context}"]
+== Understanding the Prometheus trigger
```
We should add a paragraph or subsection here, in case users would like to scale apps based on Prometheus, that is part of OpenShift Monitoring -> specific setup is required here. I described it here: https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring

mburke5678 replied:
@zroubalik 
Can people use the Monitoring docs here?
https://docs.openshift.com/container-platform/4.10/monitoring/configuring-the-monitoring-stack.html#enabling-query-logging-for-thanos-querier_configuring-the-monitoring-stack

zroubalik replied:
Yeah, this is a great reference for enabling of monitoring for user workloads (this is an prerequisite).

But then they need to:
[1. create a ServiceAccount to get a token](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#2-create-a-service-account-or-reuse-existing-one-locate-assigned-token) 
[2. create a Role](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#4-create-a-role-for-reading-metric-from-thanos) 
[3. add that role to the previously created Service Account](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#5-add-the-role-for-reading-metrics-from-thanos-to-the-service-account).
[4. and finally reference the token (secret) in the TriggerAuthentication object used by Prometheus Scaler](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#3-define-triggerauthentication-with-the-service-accounts-token)

We should cover this in the docs, becuase it is not very straightforward and we should enable users to use OpenShift Monitoring Prometheus (Thanos) instance.

zroubalik replied:
A section like `How to scale based on metrics from OpenShift Monitoring`

mburke5678 replied:
See: [Configuring the custom metrics autoscaler to use Prometheus Cluster Monitoring](http://file.rdu.redhat.com/~mburke/OSDOCS-3048-custom-autoscaling-keda/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-prometheus-config_nodes-pods-autoscaling-custom)

zroubalik replied:
looking good, could you please add a link to that section in here? So users reading about Prometheus trigger could quickly check this guide?

zroubalik reviewed (commented):
I like the direction where this is going, great stuff!

mburke5678 commented on modules/nodes-pods-autoscaling-custom-about.adoc:
```diff
@@ -0,0 +1,25 @@
+
+:_content-type: CONCEPT
+[id="nodes-pods-autoscaling-custom-about_{context}"]
+= Understanding the custom metrics autoscaler
```
@zroubalik

> it is not possible to combine HPA and ScaledObject

Does this mean, if a user configures one ScaledObject/ScaledJob, they can not use HPA directly at all?

mburke5678 replied:
How is this:

> You can create only one scaled object or scaled job for each workload that you want to scale. Also, you cannot use a scaled object or scaled job and the Horizontal Pod Authoscaler (HPA) on the same workload. If you want to scale based on a custom trigger and CPU/Memory, you can create multiple triggers in the scaled object or scaled job.

mburke5678 replied:
@zroubalik We tend to use[ lower case when we talk about an API object in a general way and use the API name when referring to a specific object](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting). 

As in:

> You need to create a scaled object to use custom metrics. 
> Add the following information in a  `ScaledObject` custom resource.

zroubalik replied:
Okay! Let's keep the convetion then (even though I don't like it 🥲 )

mburke5678 commented on modules/nodes-pods-autoscaling-custom-creating-workload.adoc:
```diff
@@ -0,0 +1,130 @@
+
+:_content-type: PROCEDURE
+[id="nodes-pods-autoscaling-custom-creating-workload_{context}"]
+= Adding a custom metrics autoscaler to a workload
```
@zroubalik @weinliu Have we tested all of the parameters and behaviors described in the KEDA docs? I don't feel comfortable documenting behaviors we have not tested. 

For example, this seems like important information. But, do we know this to be true and works as stated?

> KEDA will keep track of the number of consecutive times each scaler has failed to get metrics from its source. Once that value passes the failureThreshold, instead of not propagating a metric to the HPA (the default error behaviour), the scaler will, instead, return a normalised metric using the formula:
> 
> target metric value * fallback replicas
> 
> Due to the HPA metric being of type AverageValue (see below), this will have the effect of the HPA scaling the deployment to the defined number of fallback replicas.

There are similar behaviors throughout their docs.

mburke5678 replied:
I added a link to the KEDA docs for this behavior and others. Even that I am a bit uncomfortable with.

mburke5678 commented on modules/nodes-pods-autoscaling-custom-creating-workload.adoc:
```diff
@@ -0,0 +1,130 @@
+[id="nodes-pods-autoscaling-custom-creating-workload_{context}"]
+= Adding a custom metrics autoscaler to a workload
+
+You can create a custom metrics autoscaler (CMA) for any `Deployment` or `StatefulSet` object.
```
@zroubalik What does Custom Resource mean here (from the [KEDA docs](https://keda.sh/docs/2.6/concepts/scaling-deployments/#scaling-of-custom-resources)):

>With KEDA you can scale any workload defined as any Custom Resource (for example ArgoRollout [resource](https://argoproj.github.io/argo-rollouts/)). The scaling behaves the same way as scaling for arbitrary Kubernetes Deployment or StatefulSet.

>The only constraint is that the target Custom Resource must define /scale [subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource)

zroubalik replied:
It's a Kubernetes way to extend the standard set of provided resources (think of Pod, Deployment, HPA...). Basically any new resouce that we are shipping with an operator is a CustomResource (in this case ScaledObject/ScaledJob/TriggerAuth..). 
https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

So if users are using some CustomResources of their own, they can use KEDA to scale them out (if they define /scale subresource).

mburke5678 commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,157 @@
+[id="nodes-pods-autoscaling-custom-trigger-cpu_{context}"]
+== Understanding the CPU trigger
+
+You can scale based on CPU metrics. 
```
How about:

> This trigger uses cluster metrics as the source for metrics.

mburke5678 commented:
@zroubalik Thank you for your thorough and thoughtful comments. Not just corrections and clarifications. But also things that made the docs better.

mburke5678 commented:
@weinliu Please take a look. Thank you in advance!

weinliu commented:
@mburke5678 Hi Mike, in the session of **Uninstalling the Custom Metrics Autoscaler Operator**

 http://file.rdu.redhat.com/mburke/OSDOCS-3048-custom-autoscaling-keda/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-uninstalling_nodes-pods-autoscaling-custom

We need to delete KedaController first from web console or cli, or we'll an issue: after uninstalling, deleting the namespace of openshift-keda hangs there.

Other parts lgtm

zroubalik commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,199 @@
+
+.Prerequisites
+
+* If you use a custom metrics autoscale for scaling on CPU/memory, your cluster administrator must have properly configured cluster metrics. You can use the `oc describe PodMetrics <pod-name>` command to determine if metrics are configured. If metrics are configured, the output appears similar to the following, with CPU and Memory displayed under Usage.
```
This prerequisite here doesn't make much sense (it's only valid for CPU/Memory scaler), this is Prometheus one. Enabling OpenShift Monitoring for user workload should be here instead.

zroubalik commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,199 @@
+
+:_content-type: PROCEDURE
+[id="nodes-pods-autoscaling-custom-prometheus-config_{context}"]
+= Configuring the custom metrics autoscaler to use Prometheus Cluster Monitoring
```
This should be IMHO:
```suggestion
= Configuring the custom metrics autoscaler to use OpenShift Container Platform monitoring
```
-> http://file.rdu.redhat.com/mburke/OSDOCS-3048-custom-autoscaling-keda/monitoring/monitoring-overview.html

zroubalik commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,199 @@
+[id="nodes-pods-autoscaling-custom-prometheus-config_{context}"]
+= Configuring the custom metrics autoscaler to use Prometheus Cluster Monitoring
+
+You can use the installed Prometheus Cluster Monitoring as a source for metrics for the custom metric autoscaler. However, there is some additional configuration you must perform.
```
```suggestion
You can use the installed OpenShift Container Platform monitoring as a source for metrics for the custom metric autoscaler. However, there is some additional configuration you must perform.
```

mburke5678 replied:
@zroubalik I added this to the topic also. Is this correct? We don't list this server address in the docs. I assume it's not needed for regular use.

> To use {product-title} monitoring as the soure, in the trigger/scaler, specify the `prometheus` type and use `https://thanos-querier.openshift-monitoring.svc.cluster.local:9092` as the `serverAddress`. `

zroubalik replied:
As of today the address needs to be specified, should be fixed in future release though.

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,158 @@
+
+[discrete]
+[id="nodes-pods-autoscaling-custom-trigger-cpu_{context}"]
+== Understanding the CPU trigger
```
We should note that this trigger can't be used in ScaledJobs.

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,158 @@
+
+[discrete]
+[id="nodes-pods-autoscaling-custom-trigger-memory_{context}"]
+== Understanding the memory trigger
```
We should note that this trigger can't be used in ScaledJobs.

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,116 @@
++
+[source,terminal]
+----
+oc delete crd scaledobjects.keda.k8s.io
```
These two blocks are missing the $

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,116 @@
+
+* Uninstall the community KEDA, if present in your cluster. You cannot run both KEDA and the custom metrics autoscaler on the same {product-title} cluster.
+
+* Remove the KEDA 1.x custom resource definitions by running the following commands:
```
Is this only required if they were using the community KEDA?

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,116 @@
+. On the *Install Operator* page, ensure that the *All namespaces on the cluster (default)* option
+is selected for *Installation Mode*. This installs the Operator in all namespaces.
+
+. Ensure that the *openshift-keda* namespace is selected for *Installed Namespace*.  
```
Just checking - does the user have to create this namespace first?

mburke5678 replied:
@bergerhoffer No

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,116 @@
+
+.. Navigate to *Workloads* -> *Deployments* to verify that the `custom-metrics-autoscaler-operator` deployment is running.
+
+. Optional. Verify the installation in the {product-title} CLI using the following commands:
```
* I think we use "Optional:" with a colon instead of a period.
* "{product-title} CLI" should just be "OpenShift CLI"

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,116 @@
+<3> Specifies the logging format for the Custom Metrics Autoscaler Operator log messages. The allowed values are `console` or `json`. The default is `console`.
+<4> Specifies the logging level for the Custom Metrics Autoscaler Metrics Server. The allowed values are `0` for `info` and `4` or `debug`. The default is `0`.
++
+See the link:https://github.com/kedacore/keda-olm-operator[KEDA OLM Operator] document for all available parameters.
```
Just checking if you cleared this GH link with Eric Rich?

mburke5678 replied:
@bergerhoffer 
Yes! That's what I wanted you to ask!

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,175 @@
+
+You must perform the following tasks, as described in this section:
+
+* create a service account to get a token
```
These should be capitalized

mburke5678 replied:
Need periods, too, in that case!

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,175 @@
++
+<service_account>:: Specifies the name of the service account.
++
+.Example output
```
I think this example output is indented too far. It's indented with the service_account description, but it should be left aligned with the command block above it

mburke5678 replied:
@bergerhoffer Do you know how to configure this in asciidoc?

bergerhoffer replied:
@mburke5678 I usually have to play with things like this to see what works.

I think if you put a blank line above the + above the example output, it shifts the indentation correctly.

If that doesn't work, I think you can also use `--` around things to keep the indentation right. Like this might work:

```
[source,terminal]
----
$ oc describe serviceaccount <service_account>
----
+
--
where:

<service_account>:: Specifies the name of the service account.
--
+
.Example output
[source,terminal]
```

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,166 @@
+
+[NOTE]
+====
+If Prometheus is taking metrics from the application that the custom metric autoscaler is scaling, do not set the minimum replicas to `0` in the custom resource. If there were no application pods, the custom metric autoscaler would not have any metrics to scale on.
```
```suggestion
If Prometheus is taking metrics from the application that the custom metric autoscaler is scaling, do not set the minimum replicas to `0` in the custom resource. If there are no application pods, the custom metric autoscaler does not have any metrics to scale on.
```
Does this mean the same thing?

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,166 @@
+  name: prom-scaledobject
+  namespace: my-namespace
+spec:
+
```
I'd remove the extra blank lines

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-uninstalling.adoc:
```diff
@@ -0,0 +1,101 @@
+
+[NOTE]
+====
+Due to a known issue, you should delete the `KedaController` custom resource (CR) first. If you do not specifically delete the CR, {product-title} can hang when you delete the `openshift-keda` project. If you delete the Custom Metrics Autoscaler Operator before deleting the CR, you are not able to delete the CR.
```
I'd either link the known issue BZ here, or remove "Due to a known issue"

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-uninstalling.adoc:
```diff
@@ -0,0 +1,101 @@
+.. Delete the custom metrics autoscaler CRDs: 
++
+--
+* clustertriggerauthentications.keda.sh
```
Backticks for these?

bergerhoffer reviewed (commented):
Nice job, just lots of little nits!

mburke5678 commented on nodes/pods/nodes-pods-autoscaling-custom.adoc:
```diff
@@ -0,0 +1,62 @@
+
+include::modules/nodes-pods-autoscaling-custom-trigger-auth.adoc[leveloffset=+1]
+
+.Additional references
```
Why do I keep doing that??

mburke5678 commented on nodes/pods/nodes-pods-autoscaling-custom.adoc:
```diff
@@ -0,0 +1,62 @@
+:_content-type: ASSEMBLY
+:context: nodes-pods-autoscaling-custom
+[id='nodes-pods-autoscaling-custom']
```
Nice catch. I copied that from somewhere. Gotta find out where and fix it there, too.

mburke5678 commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,166 @@
+
+Triggers, also known as scalers, provide the metrics that the Custom Metrics Autoscaler Operator uses to scale your pods. You can specify a single or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers.
+
+The following are examples of some triggers you can use with the custom metrics autoscaler (CMA). For a full list of available triggers, see the link:https://keda.sh/docs/2.7/scalers[KEDA Scalers] documentation. Because the custom metrics autoscaler is Technology Preview, Red Hat does not recommend using the triggers/scalers in production. 
```
They are further down. I will change this wording.

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-creating-workload.adoc:
```diff
@@ -0,0 +1,156 @@
+<5> Optional: Specifies the maximum number of replicas when scaling up. The default is `100`.
+<6> Optional: Specifies the minimum number of replicas when scaling down.
+<7> Optional: Specifies the interval in seconds to check each trigger on, in seconds. The default is `30`.
+<8> Optional: Specifies the number of replicas to fallback to if a scaler fails to get metrics from the source for the number of times defined by the `failureThreshold` parameter. For more information on fallback behavior, see the https://keda.sh/docs/2.7/concepts/scaling-deployments/#scaledobject-spec[KEDA documentation]. 
```
I see what you mean, but it's used as a verb here (and is shown as "fall back" verb in the ISG)

mburke5678 replied:
@bergerhoffer Ugh. I didn't see the first one! I saw only _fallback behavior_!

bergerhoffer commented on modules/nodes-pods-autoscaling-custom-creating-workload.adoc:
```diff
@@ -0,0 +1,156 @@
++
+[source,terminal]
+----
+$ oc create -f <file-name>.yaml
```
:) you could say that about a lot of problems that we catch during peer review

sferich888 commented on modules/nodes-pods-autoscaling-custom-about.adoc:
```diff
@@ -0,0 +1,30 @@
+
+The custom metrics autoscaler uses the Kubernetes-based Event Driven Autoscaler (KEDA) and is built on top of the {product-title} horizontal pod autoscaler (HPA).
+
+The Custom Metrics Autoscaler Operator scales your pods up and down based on custom, external metrics from specific applications. Your other applications continue to use other scaling methods. You can configure a list of predefined link:https://keda.sh/#scalers[scalers], also known as triggers, which are the source of events and metrics that the custom metrics autoscaler uses to determine how to  ...
```
Is Red Hat supporting all of the Scalers listed by the upstream? IE: if a customer finds a bug with one of these scalers (or the upstream adds a scaler in the middle of a 4.Y version's lifecycle) can a customer take advantage of it? Can they get patches and security updates for the scalers? 

If the answer is no; to this then we shouldn't like to the upstream here; but instead we should list what Red Hat is supporting. 

All in all I don't have an issue linking to the upstream here; but we have to clarify this paragraph with a support clause / call out that said (more or less): Red Hat provides and only supports the ability to use scalers; the source / support for each individual scaler is offered by the provider of at scaler (please see the provider for support options), etc, etc.

gauravsingh85 replied:
We will not support all scaler but only one "prometheus" . We should be make it absolutely clear that we will not support any upstream scaler but only prometheus

mburke5678 replied:
@gauravsingh85 We explicitly document the Prometheus, Kafka, CPU, and Memory triggers.  Do we want to remove all but Prometheus?

gauravsingh85 replied:
yup only prometheus .. and leave a note that more is coming ( something like that)

mburke5678 replied:
@gauravsingh85 The internal Prometheus Cluster Monitoring and external Proetheus source? Or just internal?

zroubalik replied:
CPU and Memory scaler is just a wrapper around existing HPA funcitonality, there's no harm in supporting these. But we can keep it to the next release.

sferich888 commented on modules/nodes-pods-autoscaling-custom-install.adoc:
```diff
@@ -0,0 +1,118 @@
+<3> Specifies the logging format for the Custom Metrics Autoscaler Operator log messages. The allowed values are `console` or `json`. The default is `console`.
+<4> Specifies the logging level for the Custom Metrics Autoscaler Metrics Server. The allowed values are `0` for `info` and `4` or `debug`. The default is `0`.
++
+See the link:https://github.com/kedacore/keda-olm-operator[KEDA OLM Operator] document for all available parameters.
```
We shouldn't link to the upstream for configuration advice; the parameters we support with the operator should be documented in the OCP docs. 

I don't think this link meets the threshold or criteria for when/why we want to use a github link.

sferich888 commented on modules/nodes-pods-autoscaling-custom-trigger-auth.adoc:
```diff
@@ -0,0 +1,127 @@
+
+A trigger authentication allows you to include authentication information in a scaled object or a scaled job that can be used by the associated containers. You can use trigger authentications to pass {product-title} secrets, platform-native pod authentication mechanisms, environment variables, and so on.   
+
+Some triggers, or scalers, allow you to set specific authentication parameters by using a trigger authentication. For information, see each trigger definition in the link:https://keda.sh/docs/2.7/scalers/[KEDA Scalers] documentation.
```
Is Red Hat supporting all of the Scalers listed by the upstream? IE: if a customer finds a bug with one of these scalers (or the upstream adds a scaler in the middle of a 4.Y version's lifecycle) can a customer take advantage of it? Can they get patches and security updates for the scalers?

If the answer is no; to this then we shouldn't like to the upstream here; but instead we should list what Red Hat is supporting.

All in all I don't have an issue linking to the upstream here; but we have to clarify this paragraph with a support clause / call out that said (more or less): Red Hat provides and only supports the ability to use scalers; the source / support for each individual scaler is offered by the provider of at scaler (please see the provider for support options), etc, etc.

sferich888 commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,162 @@
+
+Triggers, also known as scalers, provide the metrics that the Custom Metrics Autoscaler Operator uses to scale your pods. You can specify a single trigger or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers.
+
+This section contains examples of some triggers you can use with the custom metrics autoscaler (CMA). For a full list of available triggers, see the link:https://keda.sh/docs/2.7/scalers[KEDA Scalers] documentation. Because the custom metrics autoscaler is Technology Preview, Red Hat does not recommend using the triggers/scalers in production. 
```
Is Red Hat supporting all of the Scalers listed by the upstream? IE: if a customer finds a bug with one of these scalers (or the upstream adds a scaler in the middle of a 4.Y version's lifecycle) can a customer take advantage of it? Can they get patches and security updates for the scalers?

If the answer is no; to this then we shouldn't like to the upstream here; but instead we should list what Red Hat is supporting.

All in all I don't have an issue linking to the upstream here; but we have to clarify this paragraph with a support clause / call out that said (more or less): Red Hat provides and only supports the ability to use scalers; the source / support for each individual scaler is offered by the provider of at scaler (please see the provider for support options), etc, etc.

sferich888 commented on modules/nodes-pods-autoscaling-custom-creating-workload.adoc:
```diff
@@ -0,0 +1,158 @@
+<6> Optional: Specifies the maximum number of replicas when scaling up. The default is `100`.
+<7> Optional: Specifies the minimum number of replicas when scaling down.
+<8> Optional: Specifies the interval in seconds to check each trigger on. The default is `30`.
+<9> Optional: Specifies the number of replicas to fall back to if a scaler fails to get metrics from the source for the number of times defined by the `failureThreshold` parameter. For more information on fallback behavior, see the https://keda.sh/docs/2.7/concepts/scaling-deployments/#scaledobject-spec[KEDA documentation]. 
```
If we were to use a link it should be https://keda.sh/docs/2.7/concepts/scaling-deployments/#fallback not the one currently in use. 

That said; I think the information that the upstream has could be/should be down streamed (thus I don't understand the need for the link. 

This information is critical to understanding a key feature / capability of the offering; and should be included in the docs for the functionality (IMO). As a result I am not supportive of a link here.

mburke5678 replied:
@gauravsingh85 @joelsmith PTAL at Eric's concerns.

weinliu replied:
@mburke5678 fallback is in the part of APPENDIX, so I did not include it in the test scope. https://github.com/zroubalik/keda-openshift-examples/tree/main/kafka/amqstreams

mburke5678 replied:
@weinliu I removed `fallback` from the docs

sferich888 reviewed (changes requested):
I don't think we should be using any of the links to upstream information here; Everything contained in the upstream links could be down streamed docs (IMO) and should be to properly convey the level of support we are offering to customers.

mburke5678 commented:
@gauravsingh85 @joelsmith PTAL at Eric's concerns.

mburke5678 commented:
@sferich888 @gauravsingh85 

Removed/hid all links to KEDA docs.
Hid Kafka, CPU, Memory trigger examples
Added wording that we only support Prometheus trigger
Changed/hid wording that user can specify multiple triggers 

See 411eb2ae16ba898077d7e3ad67418fef415b910c and https://github.com/openshift/openshift-docs/pull/42918/commits/d7ef086a263e87c7aacc5a7359ed8159ad8453da 

cc @joelsmith @zroubalik

zroubalik commented on modules/nodes-pods-autoscaling-custom-about.adoc:
```diff
@@ -0,0 +1,37 @@
+
+The custom metrics autoscaler uses the Kubernetes-based Event Driven Autoscaler (KEDA) and is built on top of the {product-title} horizontal pod autoscaler (HPA).
+
+The Custom Metrics Autoscaler Operator scales your pods up and down based on custom, external metrics from specific applications. Your other applications continue to use other scaling methods. You configure _triggers_, also known as scalers, which are the source of events and metrics that the custom metrics autoscaler uses to determine how to scale. The custom metrics autoscaler uses a metrics AP ...
```
`{product-title} curently supports only the Prometheus trigger` -> this is rendered as  `OpenShift Container Platform curently supports only the Prometheus trigger`

But should probably be:
`The the custom metrics autoscaler curently supports only the Prometheus trigger`

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,168 @@
+
+[NOTE]
+====
+{product-title} curently supports only the Prometheus trigger, which can use the installed {product-title} monitoring or an external Prometheus server as the metrics source.  
```
`{product-title} curently supports only the Prometheus trigger` -> this is rendered as  `OpenShift Container Platform curently supports only the Prometheus trigger`

But should probably be:
`The the custom metrics autoscaler curently supports only the Prometheus trigger`

zroubalik commented on nodes/pods/nodes-pods-autoscaling-custom.adoc:
```diff
@@ -0,0 +1,68 @@
+
+[NOTE]
+====
+{product-title} curently supports only the Prometheus trigger, which can use the installed {product-title} monitoring or an external Prometheus server as the metrics source.  
```
`{product-title} curently supports only the Prometheus trigger` -> this is rendered as  `OpenShift Container Platform curently supports only the Prometheus trigger`

But should probably be:
`The the custom metrics autoscaler curently supports only the Prometheus trigger`

zroubalik commented on modules/nodes-pods-autoscaling-custom-trigger.adoc:
```diff
@@ -0,0 +1,168 @@
+----
+<1> Specifies Prometheus as the scaler/trigger type.
+<2> Specifies the address of the Prometheus server. This example uses  {product-title} monitoring.
+<3> Optional: Specifies the namespace of the object you want to scale.
```
```suggestion
<3> Optional: Specifies the namespace of the object you want to scale. This property must be set if {product-title} monitoring Prometheus instance is being used.
```

zroubalik commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,175 @@
+
+.Prerequisites
+
+* {product-title} monitoring must be installed.
```
Another prerequisite is, that monitoring of user-defined workloads must be enabled in {product-title} monitoring: and paste a link here: http://file.rdu.redhat.com/mburke/OSDOCS-3048-custom-autoscaling-keda/monitoring/configuring-the-monitoring-stack.html#creating-user-defined-workload-monitoring-configmap_configuring-the-monitoring-stack

zroubalik commented on modules/nodes-pods-autoscaling-custom-prometheus-config.adoc:
```diff
@@ -0,0 +1,175 @@
+$ oc create -f <file-name>.yaml
+----
+
+You can now deploy a scaled object or scaled job to enable autoscaling for your application, as described in the following sections. To use {product-title} monitoring as the source, in the trigger, or scaler, specify the `prometheus` type and use `\https://thanos-querier.openshift-monitoring.svc.cluster.local:9092` as the `serverAddress`.
```
Would be nice if we can show here the actual ScaledObject and highlight that the user needs to use `\https://thanos-querier.openshift-monitoring.svc.cluster.local:9092` as the `serverAddress` and that `namespace` property in the trigger configuration must be set to the namespace where the workload is deployed.
https://github.com/zroubalik/keda-openshift-examples/blob/main/prometheus/ocp-monitoring/scaledobject.yaml

mburke5678 commented:
/cherrypick enterprise-4.10

mburke5678 commented:
/cherrypick enterprise-4.11
